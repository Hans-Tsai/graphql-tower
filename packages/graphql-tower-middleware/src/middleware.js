import _ from 'lodash';
import { UnauthorizedError } from 'graphql-tower-errors';

export function authentication(payload, args, context) {
  if (!_.has(context, 'user.id')) {
    throw new UnauthorizedError('authorization header is required');
  }
}

export { authentication as default };
