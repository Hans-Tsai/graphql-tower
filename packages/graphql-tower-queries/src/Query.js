import _ from 'lodash';
import PayloadField from './PayloadField';
import GraphQLField from './GraphQLField';

export default class Query extends GraphQLField {
  static resolveArgs(...args) {
    return _.defaultsDeep(_.mapValues(this._.parentArgs, (value, key) => (
      value instanceof PayloadField ? value.resolveValue(...args, key) : value
    )), args[1]);
  }

  static async preResolve(payload, client, context, info) {
    const { resolveArgs, middleware, resolve } = this.constructor;
    const args = resolveArgs.call(this, payload, client, context, info);
    await middleware.call(this, payload, args, context, info);
    return resolve.call(this, payload, args, context, info);
  }

  constructor(parentArgs) {
    super();

    this._.parentArgs = parentArgs;

    Object.defineProperty(this, 'resolve', {
      enumerable: true,
      set: (resolve) => { this._.resolve = resolve; },
      get: () => this.constructor.preResolve.bind(this),
    });
  }
}
