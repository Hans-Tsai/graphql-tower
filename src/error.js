import { GraphQLError } from 'graphql/error';

export class ForbiddenError extends GraphQLError {
  name = 'ForbiddenError';
}

export class NotFoundError extends GraphQLError {
  name = 'NotFoundError';
}

export class UnauthorizedError extends GraphQLError {
  name = 'UnauthorizedError';
}
