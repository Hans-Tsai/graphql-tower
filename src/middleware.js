import _ from 'lodash';
import { UnauthorizedError } from './error';

export function authentication(payload, args, context) {
  const user = _.get(context, 'user', {});

  if (!(_.isPlainObject(user) && user.id > 0)) {
    throw new UnauthorizedError('authorization header is required');
  }
}

export { authentication as default };
