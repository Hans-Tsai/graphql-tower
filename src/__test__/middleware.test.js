import faker from 'faker';
import { GraphQLInt } from 'graphql';
import { authentication } from '../middleware';
import Query from '../query';
import { UnauthorizedError } from '../error';

describe('afterware', () => {
  it('pagination', async () => {
    const id = faker.random.number();
    const resolve = jest.fn(() => faker.helpers.userCard());
    const middleware = jest.fn();

    const QueryAuthentication = class extends Query {
      type = GraphQLInt;
      resolve = resolve;
      middleware = [middleware, authentication];
    };
    const query = new QueryAuthentication();
    expect(query).toMatchSnapshot();

    let error;

    error = null;
    try { await query.resolve(); } catch (e) { error = e; }
    expect(error).toEqual(new UnauthorizedError('authorization header is required'));

    error = null;
    try { await query.resolve({}, {}, faker.random.word()); } catch (e) { error = e; }
    expect(error).toEqual(new UnauthorizedError('authorization header is required'));

    error = null;
    try { await query.resolve({}, {}, { user: { id } }, {}); } catch (e) { error = e; }
    expect(error).toBeNull();
  });
});
