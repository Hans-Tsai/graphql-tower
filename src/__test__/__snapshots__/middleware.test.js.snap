// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`afterware pagination 1`] = `
QueryAuthentication {
  "_": Object {
    "afterware": Array [],
    "middleware": Array [
      [Function],
      [Function],
    ],
    "resolve": [Function],
  },
  "args": Object {},
  "type": "Int",
}
`;
